pkg/metrics/metrics.go

```diff
@@ -16,45 +16,78 @@ package metrics
 
 import (
 	"context"
-	"net/http"
-	"strconv"
+	"fmt"
 
 	"github.com/operator-framework/operator-sdk/pkg/k8sutil"
 
-	"github.com/prometheus/client_golang/prometheus/promhttp"
 	v1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
 	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
 )
 
 var log = logf.Log.WithName("metrics")
 
-// ExposeMetricsPort generate a Kubernetes Service to expose metrics port
-func ExposeMetricsPort() *v1.Service {
-	http.Handle("/"+k8sutil.PrometheusMetricsPortName, promhttp.Handler())
-	go http.ListenAndServe(":"+strconv.Itoa(k8sutil.PrometheusMetricsPort), nil)
+// PrometheusPortName defines the port name used in kubernetes deployment and service resources
+const PrometheusPortName = "metrics"
 
-	service, err := k8sutil.InitOperatorService()
+// ExposeMetricsPort creates a Kubernetes Service to expose the passed metrics port.
+func ExposeMetricsPort(port int32, mgr manager.Manager) (*v1.Service, error) {
+	// We do not need to check the validity of the port, as controller-runtime
+	// would error out and we would never get to this stage.
+	s, err := initOperatorService(port, PrometheusPortName)
 	if err != nil {
-		log.Error(err, "Failed to initialize service object for operator metrics")
-		return nil
+		return nil, fmt.Errorf("failed to initialize service object for metrics: %v", err)
 	}
-	kubeconfig, err := config.GetConfig()
+	err = createService(mgr, s)
+
+	return s, nil
```

Comment:
If the service already exists, we return the one returned by `initOperatorService`, not the existing one. Just making sure that's what we want to happen.

Reply:
Actually, there is a missing error there in general 🤦‍♀️ 

> Just making sure that's what we want to happen.

Yes not sure there, do you think requesting a Service would be best and return that, or just always error out and not return a service and by that removing the exist check?

Reply:
This is not a straightforward answer due to the fact that an operator might be running in a deployment with leader election enabled. I think I've opened up a can of worms.

So there are two related concerns here:
1. Which pods are selected by the Service's selector? Right now it looks like all pods in the deployment? I'm not totally familiar with how the metrics work, but I'd imagine that we want Prometheus to scrape only the leader's metrics, right?
2. What's the lifecycle of the Service? Does it live for the duration of the leader pod or of the operator deployment?

I think we need to answer those questions to make sure we get the logic correct in `ExposeMetricsPort`.

Reply:
> Which pods are selected by the Service's selector? Right now it looks like all pods in the deployment? 

Yes since the service uses the same selector as the Deployment `name=<operator-name>` it will expose the ports for all Deployment pods.
And it's alright if we scrape the metrics for all non-leader pods as well. If a leader pod steps down for some reason it can have different metrics from the new leader (e.g number of reconciles) which is worth exporting.

> What's the lifecycle of the Service? Does it live for the duration of the leader pod or of the operator deployment?

I think it should live for the duration of the deployment. Recreating the service every time there's a new leader doesn't make much sense if it's going to be selecting all replicas. So if a service does not exist, the leader pod should create it with the ownerref set to the Deployement that owns it.

> do you think requesting a Service would be best and return that, or just always error out and not return a service and by that removing the exist check?

I think `ExposeMetricsPort()` should return the actual service that exists. Either it get's created by the pod if it doesn't exist, or it gets and returns the service created by a previous pod.

Although it's worth considering any drawbacks to tying the service lifecycle to the operator Deployment if any. I can't think of any right now though.

/cc @shawn-hurley

Reply:
> Yes since the service uses the same selector as the Deployment name=<operator-name> it will expose the ports for all Deployment pods.
And it's alright if we scrape the metrics for all non-leader pods as well. If a leader pod steps down for some reason it can have different metrics from the new leader (e.g number of reconciles) which is worth exporting.

Will prometheus scrape each service endpoint individually or will it scrape the service and get round-robin-ed among each endpoint? If the former, then that sounds good. If the latter, won't that cause problems with prometheus (e.g. counters will jump up and down depending on which pod services the request)?

Reply:
@joelanford 

> Will prometheus scrape each service endpoint individually or will it scrape the service and get round-robin-ed among each endpoint? If the former, then that sounds good. If the latter, won't that cause problems with prometheus (e.g. counters will jump up and down depending on which pod services the request)?

Prometheus uses the service and the endpoints object the service creates and maintains for target discovery and the service for additional metadata, so no we’re not load balancing. And generally we always need to monitor all pods, whether they’re a leader or not is not relevant. Better to have metrics than not have metrics, when troubleshooting, I think the idiom is “Instrument first, ask questions later.“.

Reply:
> Which pods are selected by the Service's selector? Right now it looks like all pods in the deployment?

That is true that all are selected, due to the deployment being the selector, but the manager is not started so the metrics are not setup in the controller-runtime due to the leader lock. 
@hasbro17 @joelanford we have two options:
1. move the lock https://github.com/operator-framework/operator-sdk/blob/master/pkg/scaffold/cmd.go#L92 further down after the https://github.com/operator-framework/operator-sdk/blob/master/pkg/scaffold/cmd.go#L99 manager setup? And with that move the metrics setup as well.
2. Just expose metrics for the current or previous leaders.

I would go with 1. due to my previous comment, but I am not sure if this breaks anything for other components?

Reply:
I agree with option 1.

Reply:
I agree with option 1.

Reply:
Did some testing and 1. doesn't work because metrics get served only when `Start` is called. I would like to suggest upstream to serve metrics before that as it should be independantly of `Start`. I am assuming that we should hold the lock before [that](https://github.com/operator-framework/operator-sdk/blob/master/pkg/scaffold/cmd.go#L122), right? 

They do handle leader election and serve metrics if there is a lock, but we use our own leader logic so we can't rely on that IIUC.

I guess for now so we have some metrics, I would suggest we go with option 2. and contribute upstream if they agree on serving metrics independantly. SGTY @shawn-hurley @hasbro17 @joelanford ?

Reply:
Sum up from discussion with @shawn-hurley:

1. Will look into disabling serving of metrics in controller-runtime, so we can serve the metrics they expose in operator-sdk.
2. Open issue to discuss starting serving metrics independently of `Start()` being called.

Reply:
Agreed. Option 2 sounds fine until we can work upstream or find a workaround to expose the metrics before getting the lock and calling start.

Reply:
> Agreed. Option 2 sounds fine until we can work upstream or find a workaround to expose the metrics before getting the lock and calling start.

Okay, will open an issue to make it work for all pods, not just leader when this is merged 👍 Until then we will at least have some metrics.